interceptors/logging/logging.go

```diff
@@ -37,46 +35,110 @@ var (
 	MethodTypeFieldKey   = "grpc.method_type"
 )
 
-func commonFields(kind string, typ interceptors.GRPCType, service string, method string) Fields {
+type fieldsCtxMarker struct{}
+
+var (
+	// fieldsCtxMarkerKey is the Context value marker that is used by logging middleware to read and write logging fields into context.
+	fieldsCtxMarkerKey = &fieldsCtxMarker{}
+)
+
+// Fields represents logging fields. It has to have even number of elements (pairs).
+type Fields []string
+
+func newCommonFields(kind string, c interceptors.CallMeta) Fields {
 	return Fields{
 		SystemTag[0], SystemTag[1],
 		ComponentFieldKey, kind,
-		ServiceFieldKey, service,
-		MethodFieldKey, method,
-		MethodTypeFieldKey, string(typ),
+		ServiceFieldKey, c.Service,
+		MethodFieldKey, c.Method,
+		MethodTypeFieldKey, string(c.Typ),
 	}
 }
 
-// Fields represents logging fields. It has to have even number of elements (pairs).
-type Fields []string
+func (f Fields) Iter() FieldsIter {
+	return iter{i: -2, f: f}
+}
 
-// ErrorToCode function determines the error code of an error
-// This makes using custom errors with grpc middleware easier
-type ErrorToCode func(err error) codes.Code
+type FieldsIter interface {
+	Next() bool
+	At() (k, v string)
+}
+
+type iter struct {
+	f Fields
+	i int
+}
+
+func (i iter) Next() bool {
+	if i.i >= len(i.f) {
+		return false
+	}
+
+	i.i += 2
+	return i.i < len(i.f)
+}
+
+func (i iter) At() (k, v string) {
+	if i.i < 0 || i.i >= len(i.f) {
+		return "", ""
+	}
 
-func DefaultErrorToCode(err error) codes.Code {
-	return status.Code(err)
+	if i.i+1 == len(i.f) {
+		// Non even number of elements, add empty string.
+		return i.f[i.i], ""
+	}
+	return i.f[i.i], i.f[i.i+1]
 }
 
-// Decider function defines rules for suppressing any interceptor logs
-type Decider func(fullMethodName string, err error) Decision
+// AppendUnique returns fields which is the union of all keys with the added values having lower priority.
```

Comment:
This text confused me, but after reading the source, I think I'd reword this like this:
```suggestion
// AppendUnique returns fields which is the union of all keys. Any keys that already exist in the log fields will take precedence over duplicates in add.
```